quint/mempool/v0/mempoolv0.qnt

```diff
@@ -0,0 +1,558 @@
+        _error' = { node: noNode, error: noError },
+        mempoolHeight' = NodeIds.mapBy(_ => 0),
+        History::init,
+        Network::init,
+        Chain::init,
+        ABCI::init,
+    }
+    
+    // Validate an incoming transaction by sending an ABCI request to the
+    // application.
+    // [CListMempool.CheckTx]: https://github.com/CometBFT/cometbft/blob/5a8bd742619c08e997e70bc2bbb74650d25a141a/mempool/clist_mempool.go#L202
+    //-- __subaction__: this action is not part of the next-state relation
+    action CheckTx(node: NodeId, tx: Tx, senderId: NodeId): bool = all {
+        if (node.mempoolIsFull()) all {
+            node.setError(MempoolIsFullError),
+            mempool' = mempool,
+            cache' = cache,
+            ABCI::unchanged,
+        } else if (tx.in(node.Cache())) all {
+            // Record new sender for the tx we've already seen.
+            // Note it's possible a tx is still in the cache but no longer in the mempool
+            // (eg. after committing a block, txs are removed from mempool but not cache),
+            // so we only record the sender for txs still in the mempool.
+            node.addSender(tx, senderId),
+            node.setError(TxInCacheError),
+            cache' = cache,
+            ABCI::unchanged,
+        } else all {
+            node.addToCache(tx),
+            node.ABCI::sendRequestNewCheckTx(tx, senderId, node.MempoolHeight() + 1),
+            node.setError(noError),
+            mempool' = mempool,
+        },
+        mempoolHeight' = mempoolHeight,
+    }
+
+    // Callback function to process a CheckTx response of type "New".
+    // Note that tx and sender are arguments to CListMempool.resCbFirstTime.
+    // [CListMempool.resCbFirstTime]: https://github.com/CometBFT/cometbft/blob/6498d67efdf0a539e3ca0dc3e4a5d7cb79878bb2/mempool/clist_mempool.go#L369
+    action ReceiveCheckTxResponse(node) = all {
+        node.setStep("ReceiveCheckTxResponse", Map()),
+        require(node.ABCI::checkTxRequests().nonEmpty()),
+        nondet req = oneOf(node.ABCI::checkTxRequests()) 
+        val resp = node.ABCI::responseFor(req)
+        all {
+            node.ABCI::removeRequest(req),
+            if (resp.err == noError) {
+                if (node.mempoolIsFull()) all {
+                    node.forceRemoveFromCache(req.tx), // There may be enough space for tx later.
+                    node.setError(MempoolIsFullError),
+                    mempool' = mempool,
+                    History::unchanged,
+                } else all { 
+                    require(not(node.inMempool(req.tx))),
+                    val senderId = node.ABCI::senderFor(req)
+                    val h = node.ABCI::heightFor(req) // height at which tx was validated
+                    node.addToMempool(req.tx, h, senderId),
+                    node.setError(noError),
+                    node.History::recordMempoolTx(req.tx),
+                    cache' = cache,
```

Comment:
Since we are all still figuring out Quint, here is a comment on usage of Quint. It would be interesting to get your opinion on: I understand that this spec is supposed to be targeted for MBT, that is, it should be efficient for machines to read. I just want to point out here, that I (as a human ;-) ) find it hard to read. As far as I understand, this snippet contains

- recording of state for MBT (l 186 setstep)
- some modeling of the environment (l 187, l188, l191)
- then basically the body of the mentioned function (in clist_mempool.go) but again with some lines required for modeling (e.h., the ones about History).

However it is hard to grasp syntactically what is doing what. Many lines start with "`node.`":
- `node.setStep("ReceiveCheckTxResponse", Map())` is for bookkeeping within the spec
- `node.ABCI::checkTxRequests().nonEmpty()` is for modeling the environment (north of ABCI)
- `node.addToMempool(req.tx, h, senderId)` actually captures the behavior of the modeled function.

In Quint the functional layer could help to separate these concerns. I wonder whether you considered it and whether it is bad for use in MBT?

Reply:
I agree. Readability is more important than making the spec efficient for MBT. I'm also still trying to figure out what are the best practices for writing Quint. Until now I have focused mainly on making the spec work and generating traces. I will need to work on another iteration of this spec for improving how it is structured, and probably adopt a function layer as you suggest.

Reply:
> Many lines start with "node."

This is because the spec models a set of nodes and all actions are local to one of the nodes, as one would normally have in TLA+. I hope Quint can provide in the future a way to write multi-process protocols as in Pluscal (cc @konnov), where one can write:
```
--algorithm Foo { 
  variables x = 0, y = 0; // global variables
  process (Proc \in 1 .. N ) {
	variables ... // local variables
  }
}
```
Then all "node." will disappear, and updating a variable local to a node will be shorter.